### PR TITLE
Port High Scores

### PR DESCRIPTION
**Pull request by filalex77**
This change is [Reviewable](https://reviewable.io/reviews/exercism/rust/886)

### CODE REVIEW

File: exercises/high-scores/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[dependencies]
+
+[package]
+authors = ["Oleksii Filonenko <brightone@protonmail.com>"]
```
Comment (coriolinus) [inline exercises/high-scores/Cargo.toml]: Please remove this line; policy since https://github.com/exercism/rust/issues/524 has been to leave it out.

File: exercises/high-scores/example.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#[derive(Debug)]
+pub struct HighScores {
+    scores: Vec<u32>,
+}
+
+impl HighScores {
+    pub fn new(scores: &[u32]) -> Self {
+        Self {
+            scores: scores.to_vec(),
```
Comment (coriolinus) [inline exercises/high-scores/example.rs]: Instead of converting the scores input to an owned vector, this exercise could serve as a simple introduction to lifetimes by retaining a reference to the input slice. I think that direction would improve the quality of this exercise.

Necessary changes to implement that:

- update this example
- add a `.meta/hints.md` file recommending that students retain a reference to the input, and regenerate the readme
- update the topics in `config.json` to include `lifetimes`

File: exercises/high-scores/example.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    pub fn personal_best(&self) -> u32 {
+        *self.scores.iter().max().unwrap()
+    }
+
+    pub fn personal_top_three(&self) -> Vec<u32> {
+        let mut sorted = self.scores.clone();
+        sorted.sort();
+        sorted.iter().rev().take(3).cloned().collect()
```
Comment (coriolinus) [inline exercises/high-scores/example.rs]: We can't avoid the first instance of cloning the list, since we otherwise have no way to sort it. Still, we can at least avoid the second clone/collect cycle.

```suggestion
        sorted.reverse();
        sorted.truncate(3);
        sorted
```

File: exercises/high-scores/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#[derive(Debug)]
+pub struct HighScores;
+
+impl HighScores {
+    pub fn new(_: &[u32]) -> Self {
```
Comment (coriolinus) [inline exercises/high-scores/src/lib.rs]: You refer to the `scores` variable below, but you haven't set it yet.

```suggestion
    pub fn new(scores: &[u32]) -> Self {
```

File: exercises/high-scores/example.rs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+            scores: scores.to_vec(),
+        }
+    }
+
+    pub fn scores(&self) -> &[u32] {
+        &self.scores
+    }
+
+    pub fn latest(&self) -> u32 {
+        *self.scores.last().unwrap()
```
Comment (coriolinus) [inline exercises/high-scores/example.rs]: We should not encourage students to use `unwrap` unconditionally like this. We should change the API to make the struct safer.

There are at least two possible ways to change this API to improve its safety:

- change the return type of `latest` and `personal_best` to `Option<u32>`
- change the return type of `new` to `Option<Self>`, returning None when the input list is empty

I do not have a preference between those two options. However, we should definitely add an appropriate test in which the input slice is empty to ensure that the code doesn't panic, as it would with `unwrap`.

File: exercises/high-scores/Cargo.toml (hunk 1 above)
Comment (filalex77) [reply exercises/high-scores/Cargo.toml]: Okay, sure. `exercise generate` put it there for me. Maybe it should get removed automatically?

File: exercises/high-scores/example.rs (hunk 2 above)
Comment (filalex77) [reply exercises/high-scores/example.rs]: I did it with lifetimes first, and then reverted. On the second look, it's probably not that difficult with this exercise. I'll update to using references.

File: exercises/high-scores/example.rs (hunk 3 above)
Comment (filalex77) [reply exercises/high-scores/example.rs]: Truncate is coll, didn't know about it. Thanks!

File: exercises/high-scores/src/lib.rs (hunk 4 above)
Comment (filalex77) [reply exercises/high-scores/src/lib.rs]: Oops, saw the "use more placeholders" recommendation and forgot that it's used. Good catch!

File: exercises/high-scores/example.rs (hunk 5 above)
Comment (filalex77) [reply exercises/high-scores/example.rs]: I think the first variant is a bit clearer. And the test is a good idea.

File: exercises/high-scores/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+[package]
+edition = "2018"
+name = "high-scores"
+version = "4.0.0"
```
Comment (petertseng) [inline exercises/high-scores/Cargo.toml]: just confirming that I looked at this and it matches the corresponding problem-specifications version :+1: . I haven't done a full review yet, but if I don't get to it in time I trust other maintainers' judgment, so no need to wait for me.

File: exercises/high-scores/tests/high-scores.rs [hunk 7]
```diff
@@ -11,14 +11,21 @@ fn test_list_of_scores() {
 #[ignore]
 fn test_latest_score() {
     let high_scores = HighScores::new(&[100, 0, 90, 30]);
-    assert_eq!(high_scores.latest(), 30);
+    assert_eq!(high_scores.latest(), Some(30));
 }
 
 #[test]
 #[ignore]
-fn test_personal_best() {
```
Comment (coriolinus) [inline exercises/high-scores/tests/high-scores.rs]: I am surprised to see that you removed `test_personal_best` entirely. While I entirely approve of the addition of `test_latest_score_empty`, `test_personal_best_empty`, and `test_personal_top_three_empty`, I think we should not remove this canonical test case.

Comment (filalex77) [reply exercises/high-scores/tests/high-scores.rs]: I'm pretty sure it was by accident. Thanks a lot!

File: config.json [hunk 8]
```diff
@@ -1123,6 +1123,18 @@
       "difficulty": 0,
       "topics": null,
       "deprecated": true
+    },
+    {
+      "slug": "high-scores",
```
Comment (petertseng) [inline config.json]: I know that we currently have no exercise ordering check in the tests right now because https://github.com/exercism/rust/pull/804#issuecomment-470197811 . But maybe it makes sense to move high-scores above the deprecated exercise `hexadecimal`?

This change is optional - feel free to merge without having made this change without seeking approval from me on whether it is OK to not make the change

Comment (coriolinus) [reply config.json]: Good point! I hadn't actually looked at where in the configuration file this had been added. 

I agree, this exercise should fall in the region after the exercise which unlocks it (which should not be null) and before the subsequent core exercise.
